### PR TITLE
fix(bubble): tooltip visibility on hover

### PR DESCRIPTION
**Pull request by nickofthyme**
## Summary

Fixes a bug in bubble chart that prevented the hovered values from correctly being active.

### Before

Screen Recording 2022-05-17 at 09 31 57 AM

### After

Screen Recording 2022-05-17 at 09 30 01 AM

## Details

The issue is that the bubble chart was being filled on the x domain and this created bad `NaN` values in the Delaunay index and excluded those values from the tooltip list.

I'm not sure what actually caused this as I don't see or think of any recent (<16m) changes to the problematic files. Maybe the `NaN` scale changes just brought the issue to light but I don't think that is what caused this.

## Issues

#1673

### Checklist

- [x] The proper **chart type** label has been added (e.g. `:xy`, `:partition`)
- [x] The proper **feature** labels have been added (e.g. `:interactions`, `:axis`)
- [x] All related issues have been linked (i.e. `closes #123`, `fixes #123`)
- [x] Unit tests have been added or updated to match the most common scenarios

### CODE REVIEW

File: packages/charts/src/chart_types/xy_chart/renderer/canvas/renderers.ts [hunk 1]
```diff
@@ -145,17 +145,6 @@ export function renderXYChartCanvas2d(
...
-            const renderer = geometriesIndex.triangulation([0, 0, width, height])?.render;
-            if (isCanvasRenderer(renderer)) {
-              ctx.beginPath();
-              ctx.translate(left, top);
-              ctx.setLineDash([5, 5]);
-              renderer(ctx);
-              ctx.lineWidth = 1;
-              ctx.strokeStyle = 'blue';
-              ctx.stroke();
-            }
```
Comment (nickofthyme) [inline packages/charts/src/chart_types/xy_chart/renderer/canvas/renderers.ts:158]: @markov00 This fixes the `debug` error.

File: packages/charts/src/chart_types/xy_chart/utils/fill_series.ts [hunk 2]
```diff
@@ -62,11 +56,6 @@ export function fillSeries(
...
 
-function isXFillNotRequired(spec: BasicSeriesSpec, groupScaleType: ScaleType, isStacked: boolean) {
-  const onlyNoFitAreaLine = (isAreaSeriesSpec(spec) || isLineSeriesSpec(spec)) && !spec.fit;
-  const onlyContinuous =
-    groupScaleType === ScaleType.Linear ||
-    groupScaleType === ScaleType.LinearBinary ||
-    groupScaleType === ScaleType.Time;
-  return onlyNoFitAreaLine && onlyContinuous && !isStacked;
+function isXFillRequired(spec: BasicSeriesSpec, isStacked: boolean): boolean {
+  return (isAreaSeriesSpec(spec) || isLineSeriesSpec(spec)) && (isStacked || !!spec.fit);
```
Comment (markov00) [inline packages/charts/src/chart_types/xy_chart/utils/fill_series.ts:60]: I don't think this is the correct way to fix the problem and to me it looks like it doesn't fully respect the original constraints making a general change on how to fill missing values.
I know is a bit convoluted with all those double negations, but the inverted representation you are doing doesn't fully cover all the aspects that where there before.

File: packages/charts/src/chart_types/xy_chart/rendering/points.ts [hunk 3]
```diff
@@ -112,7 +112,7 @@ export function renderPoints(
       const isInYDomain = yScale.isValueInDomain(valueAccessor(datum));
       indexedGeometryMap.set(pointGeometry, geometryType);
```
Comment (markov00) [inline packages/charts/src/chart_types/xy_chart/rendering/points.ts:113]: What you can don here instead is to check if the `datum` was filled before updating the geometry map with 
```ts
if(!isDatumFilled(datum)) indexedGeometryMap.set(pointGeometry, geometryType);
```

File: packages/charts/src/chart_types/xy_chart/renderer/canvas/renderers.ts (hunk 1 above)
Comment (markov00) [reply packages/charts/src/chart_types/xy_chart/renderer/canvas/renderers.ts:158]: more then fix it avoids the error :D, I like btw the voronoi debug, there is definitely some issue with the closures or something along that blocks that renderer.
I'm fine commenting this out until we don't fix the issue if you don't mind

**Review by markov00 — CHANGES_REQUESTED** [review]
The fix should not change the current existing fill behaviour of line/area charts.
I proposed a solution that I tested locally and that could work fine without changing the that behaviour

File: packages/charts/src/chart_types/xy_chart/renderer/canvas/renderers.ts (hunk 1 above)
Comment (nickofthyme) [reply packages/charts/src/chart_types/xy_chart/renderer/canvas/renderers.ts:158]: Oh yeah I forgot about the voronoi debug 🤦‍♂️. I'll add it back and look at the closure.

File: packages/charts/src/chart_types/xy_chart/utils/fill_series.ts (hunk 2 above)
Comment (nickofthyme) [reply packages/charts/src/chart_types/xy_chart/utils/fill_series.ts:60]: That's fine but in general why are we defaulting to fill x values? By saying `isXFillNotRequired` defines only when it is skippable, that way everytime we add new chart types, we would need to update this fill it is not needed.

Comment (markov00) [reply packages/charts/src/chart_types/xy_chart/utils/fill_series.ts]: I've reverted back this to the current master, this PR https://github.com/elastic/elastic-charts/pull/1688 will take care of fixing this correctly

File: packages/charts/src/chart_types/xy_chart/utils/fill_series.ts [hunk 4]
```diff
@@ -62,11 +61,11 @@ export function fillSeries(
   });
 }
 
-function isXFillNotRequired(spec: BasicSeriesSpec, groupScaleType: ScaleType, isStacked: boolean) {
-  const onlyNoFitAreaLine = (isAreaSeriesSpec(spec) || isLineSeriesSpec(spec)) && !spec.fit;
+function isXFillRequired(spec: BasicSeriesSpec, groupScaleType: ScaleType, isStacked: boolean): boolean {
```
[the remark this answers is not in the document]
Comment (markov00) [reply packages/charts/src/chart_types/xy_chart/utils/fill_series.ts]: I'm going to fix this in a different PR: https://github.com/elastic/elastic-charts/pull/1688

**Review by markov00 — APPROVED** [review]
I've refactored the fix to push the solution more into the spatial index context rather than an an higher level.
Please @nickofthyme check my comments on this page and review the changes

File: packages/charts/src/chart_types/xy_chart/renderer/canvas/renderers.ts [hunk 5]
```diff
@@ -145,13 +145,12 @@ export function renderXYChartCanvas2d(
...
-
-            const renderer = geometriesIndex.triangulation([0, 0, width, height])?.render;
-            if (isCanvasRenderer(renderer)) {
+            const voronoi = geometriesIndex.triangulation([0, 0, width, height]);
+            if (voronoi) {
               ctx.beginPath();
               ctx.translate(left, top);
               ctx.setLineDash([5, 5]);
-              renderer(ctx);
+              voronoi.render(ctx);
```
Comment (nickofthyme) [inline packages/charts/src/chart_types/xy_chart/renderer/canvas/renderers.ts:153]: 👍🏼

File: packages/charts/src/chart_types/xy_chart/utils/indexed_geometry_spatial_map.ts [hunk 6]
```diff
@@ -42,15 +42,17 @@ export class IndexedGeometrySpatialMap {
   set(points: PointGeometry[]) {
     this.maxRadius = Math.max(this.maxRadius, ...points.map(({ radius }) => radius));
     const { pointGeometries } = this;
-    points.forEach((p) => pointGeometries.push(p));
+    points.forEach((p) => {
+      if (isFiniteNumber(p.y)) pointGeometries.push(p);
+    });
```
Comment (nickofthyme) [inline packages/charts/src/chart_types/xy_chart/utils/indexed_geometry_spatial_map.ts:47]: Nice!

File: packages/charts/src/chart_types/xy_chart/rendering/points.ts [hunk 7]
```diff
@@ -112,7 +112,7 @@ export function renderPoints(
...
       // use the geometry only if the yDatum in contained in the current yScale domain
-      if (y !== null && yDefined(datum, valueAccessor) && isInYDomain && !isDatumFilled(datum)) {
+      if (isFiniteNumber(y) && yDefined(datum, valueAccessor) && isInYDomain && !isDatumFilled(datum)) {
```
Comment (nickofthyme) [inline packages/charts/src/chart_types/xy_chart/rendering/points.ts:115]: I don't think this check is necessary at all cuz canvas just ignores `NaN` values.

```suggestion
      if (yDefined(datum, valueAccessor) && isInYDomain && !isDatumFilled(datum)) {
```

**Review by nickofthyme — COMMENTED** [review]
Can't approve my own PR but changes all look good.

**Review by markov00 — APPROVED** [review]

**Pull request merged**